### PR TITLE
UCS/MEMORY: Add new type for host memory allocated by CUDA

### PR DESCRIPTION
**Pull request by SeyedMir**
## What
Add a new UCS memory type to distinguish between host memory that is allocated by CUDA and host memory that is not allocated by CUDA.

## Why ?
Host memory that is allocated by CUDA will be released by CUDA too. So, it is managed by CUDA and we need to distinguish it from host memory that is not managed by CUDA.

### CODE REVIEW

File: src/uct/cuda/base/cuda_md.c [hunk 1]
```diff
@@ -111,12 +111,17 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_mem_query,
...
+             (cuda_mem_mype != CU_MEMORYTYPE_HOST))) {
             /* pointer not recognized */
             return UCS_ERR_INVALID_ADDR;
         }
 
-        if (is_managed) {
+        if (cuda_mem_mtype == CU_MEMORYTYPE_HOST) {
+            mem_type = UCS_MEMORY_TYPE_CUDA_HOST;
+        }
+        else if (is_managed) {
```
Comment (dmitrygx) [inline src/uct/cuda/base/cuda_md.c:124]: put on the same line with `}` from `if` block

**Review by dmitrygx — CHANGES_REQUESTED** [review]

**Comment by Akshay-Venkatesh** [comment]
@SeyedMir You would also need to add CUDA_HOST to the list of memory types that cuda_copy transport can identify here https://github.com/openucx/ucx/blob/de000dab16c18249b9af2bbfecbd29321ec3ecee/src/uct/cuda/cuda_copy/cuda_copy_md.c#L37
I don't think that's been added.

**Comment by SeyedMir** [comment]
@Akshay-Venkatesh Should I add CUDA_HOST to the list of memory types that cuda_copy transport can register too? I understand that CUDA_HOST is already registered, so I'm thinking we should not need that.

**Comment by shamisp** [comment]
@SeyedMir  - Thanks for the patch. What is your affiliation ? thanks

**Comment by SeyedMir** [comment]
@shamisp I'm with NVIDIA.

**Comment by yosefe** [comment]
Do we need to distinguish CUDA_HOST from HOST just for memory release purpose?
how is it different than host memory allocated by malloc() vs. host memory allocated by mmap()?

**Comment by bureddy** [comment]
Agree. we should not add need mem type unless we need to differentiate it from HOST memory in the protocols.
just adding to the list may also consume resources like mem_type_ep

**Comment by SeyedMir** [comment]
@yosefe @bureddy Agreed. I had a use case in mind for which I thought we need to distinguish between CUDA_HOST and HOST memory types. It is not necessary.

**Pull request closed without merging**